src/referral.cairo

```diff
@@ -0,0 +1,202 @@
+
+    // dispatchers
+    use referral::token::erc20::{ IERC20Dispatcher, IERC20DispatcherTrait };
+
+    struct Storage {
+        sponsor_balance: LegacyMap::<ContractAddress, u256>,
+        sponsor_comm: LegacyMap::<ContractAddress, u256>,
+        default_comm: u256,
+        min_claim: u256,
+        naming_contract: ContractAddress,
+        eth_contract: ContractAddress,
+    }
+
+    //
+    // Events
+    //
+
+    #[event]
+    fn on_claim(timestamp: u64, amount: u256, sponsor_addr: ContractAddress, ) {}
+
+    #[event]
+    fn on_commission(timestamp: u64, amount: u256, sponsor_addr: ContractAddress, caller: ContractAddress, ) {}
+
+
+    //
+    // Constructor
+    //
+
+    #[constructor]
+    fn constructor(
+        admin: ContractAddress, 
+        naming_addr: ContractAddress,
+        eth_addr: ContractAddress,
+        min_claim_amount: u256, 
+        share: u256
+    ) {
+        initializer(:admin);
+        naming_contract::write(naming_addr);
+        eth_contract::write(eth_addr);
+        min_claim::write(min_claim_amount);
+        default_comm::write(share);
+    }
+    
+    //
+    // View
+    //
+
+    #[view]
+    fn get_balance(sponsor_addr: ContractAddress) -> u256 {
+        sponsor_balance::read(sponsor_addr)
+    }
+
+    //
+    // External
+    //
+
+    #[external]
+    fn claim(amount: u256) {
+        assert(amount >= min_claim::read(), 'Amount is too low');
+        let sponsor_addr = get_caller_address();
+        let balance = sponsor_balance::read(sponsor_addr);
+        assert(balance >= amount, 'Amount greater than balance');
+
+        let contract_addr = get_contract_address();
+        let ERC20 = IERC20Dispatcher { contract_address: eth_contract::read() };
+        let contract_balance = ERC20.balance_of(contract_addr);
+        assert(contract_balance >= amount, 'Contract balance too low');
+        ERC20.transferFrom(sender: contract_addr, recipient: sponsor_addr, amount: amount);
+        sponsor_balance::write(sponsor_addr, balance - amount);
+
+        on_claim(get_block_timestamp(), amount, sponsor_addr);
+    }
+
+    #[external]
+    fn add_commission(amount: u256, sponsor_addr: ContractAddress) {
+        let caller = get_caller_address();
+        assert(caller == naming_contract::read(), 'Caller not naming contract');
+
+        // Calculate commission
+        let mut share = sponsor_comm::read(sponsor_addr);
+        if share == (u256 { low: 0, high: 0 }) {
+            share = default_comm::read();
+        }
+        // u256_is_zero is not accepted yet
+        // let share = match integer::u256_is_zero(sponsor_comm::read(sponsor_addr)) {
+        //     zeroable::IsZeroResult::Zero(()) => default_comm::read(),
+        //     zeroable::IsZeroResult::NonZero(x) => sponsor_comm::read(sponsor_addr),
+        // };
+
+        // todo: update to use u256_safe_divmod when we can
+        let comm = (amount.low * share.low) / 100_u128;
```

Comment:
This works because amount and share don't use all the 128 bits. Please add a `// warning: make sure to check for overflow` in case we forget in the future.